File: blog.rb
#!/usr/bin/ruby

# blog.rb
# implements functionality found in Note 2.3 (pg 69) of Distributed
# Systems

require 'rinda/rinda'

# each message for the blog is a 3-tuple <string, string, string>
# where the first string names the poster
# where the second string names the topic
# where the third string is the content

port = 61676
uri = "druby://localhost:#{port}"
DRb.start_service

# TupleSpaceProxy allows remote TS to appear as local to this process
# DRbObject.new(obj, uri) creates a new remote object stub, where the first
# argument is a LOCAL object we want to stub
blog = Rinda::TupleSpaceProxy.new(DRbObject.new(nil, uri))

# bob's blog
blog.write(["bob", "distsys", "I am studying chap 2"])
blog.write(["bob", "distsys", "The linda example's pretty simple"])
blog.write(["bob", "gtcn", "Cool book!"])

# alice's blog
blog.write(["alice", "gtcn", "This graph theory stuff is not easy"])
blog.write(["alice", "distsys", "I like systems more than graphs"])

# chuck reads both the blogs
t1 = blog.read(["bob", "distsys", String])
t2 = blog.read(["alice", "gtcn", String])
t3 = blog.read(["bob", "gtcn", String])

# '#{ruby-code}' is ruby's syntax for string interpolation
puts "#{t1}"
puts "#{t2}"
puts "#{t3}"

File: blog.py
#!/usr/bin/python

# blog.py

# connects to an XML-RPC adapter server written in Ruby to interface
# with Ruby's Rinda tuplespace

from xmlrpc.client import ServerProxy, Error

with ServerProxy("http://localhost:8000") as proxy:
    try:
        # bob's blog
        proxy.microblog._out(["bob", "distsys", "I am studying chap 2"])
        proxy.microblog._out([
            "bob", "distsys", "The linda example's pretty simple"])
        proxy.microblog._out(["bob", "gtcn", "Cool book!"])

        # alice's blog
        proxy.microblog._out([
            "alice", "gtcn", "This graph theory stuff is not easy"])
        proxy.microblog._out(["alice", "distsys",
                              "I like systems more than graphs"])

        # chuck reads both blogs
        print(proxy.microblog._rd(["bob", "distsys", {"class" : "String"}]))
        print(proxy.microblog._rd(["alice", "gtcn", {"class" : "String"}]))
        print(proxy.microblog._rd(["bob", "gtcn", {"class" : "String"}]))
    except Error as e:
        print(f'Error: {e}')
